Add serialization logic for complex numbers

anjali411 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #51145 Support complex number list in JIT
* **#50885 Add serialization logic for complex numbers**

Differential Revision: [D26094906](https://our.internmc.facebook.com/intern/diff/D26094906)

anjali411 committed:
Add serialization logic for complex numbers

[ghstack-poisoned]

anjali411 committed:
Update on "Add serialization logic for complex numbers"




[ghstack-poisoned]

anjali411 committed:
Update on "Add serialization logic for complex numbers"




[ghstack-poisoned]

anjali411 committed:
Update on "Add serialization logic for complex numbers"




[ghstack-poisoned]

anjali411 committed:
Update on "Add serialization logic for complex numbers"




[ghstack-poisoned]

anjali411 committed:
Update on "Add serialization logic for complex numbers"




[ghstack-poisoned]

anjali411 committed:
Update on "Add serialization logic for complex numbers"




[ghstack-poisoned]

SplitInfinity commented on test/test_complex.py:
```diff
@@ -18,6 +19,22 @@ def fn(a: complex):
+                return b
+
+        scripted = torch.jit.script(ComplexModule())
+        buffer = io.BytesIO()
+        torch.jit.save(scripted, buffer)
+        buffer.seek(0)
+        loaded = torch.jit.load(buffer)
```
I think you can use `self.getExportImportCopy` here, which will save and load from a buffer for you.

SplitInfinity commented on torch/csrc/jit/serialization/unpickler.cpp:
```diff
@@ -522,6 +520,8 @@ void Unpickler::readGlobal(
         elem_type = TensorType::get();
       } else if (class_name == "build_doublelist") {
         elem_type = FloatType::get();
+      } else if (class_name == "build_complexdoublelist") {
+        elem_type = ComplexDoubleType::get();
```
Do we have tests for this?

anjali411 replied:
good catch! this change shouldn't be here because complex lists are not supported yet

SplitInfinity reviewed (commented):
Nice! We should add a test for `build_complexdoublelist` or remove the deserialization logic for it. Other than that, I think this is ready to land.

anjali411 committed:
Update on "Add serialization logic for complex numbers"




[ghstack-poisoned]

anjali411 committed:
Update on "Add serialization logic for complex numbers"




[ghstack-poisoned]

anjali411 commented:
@SplitInfinity updated the PR based on the comments!

SplitInfinity commented on test/test_complex.py:
```diff
@@ -1,4 +1,5 @@
 import torch
+import io
```
```suggestion
```

SplitInfinity commented on test/test_complex.py:
```diff
@@ -12,6 +13,18 @@ def fn(a: complex):
+    def test_pickle(self):
+        class ComplexModule(torch.nn.Module):
+            def __init__(self):
+                super(ComplexModule, self).__init__()
```
```suggestion
                super().__init__()
```

SplitInfinity reviewed (approved):
Nice!

anjali411 committed:
Update on "Add serialization logic for complex numbers"


Differential Revision: [D26094906](https://our.internmc.facebook.com/intern/diff/D26094906)

[ghstack-poisoned]

mruberry commented:
Unlanding. This appears to have broken several builds. Example failure:

```
test_run_mypy [mypy.ini] - TestTypeHints
test_type_hints.py

Traceback (most recent call last):
  File "test_type_hints.py", line 171, in test_run_mypy
    self.fail(f"mypy failed: {stdout} {stderr}")
AssertionError: mypy failed: test/test_complex.py:21: error: Signature of "forward" incompatible with supertype "ScriptModule"  [override]
Found 1 error in 1 file (checked 1211 source files)
```

gchanan commented on torch/csrc/jit/serialization/unpickler.cpp:
```diff
@@ -57,6 +57,7 @@ void restoreAccurateTypeTags(const IValue& root, const TypePtr& type_tag) {
       case StorageType::Kind:
       case NumberType::Kind:
       case FloatType::Kind:
+      case ComplexDoubleType::Kind:
```
I guess this already existed but this naming looks inconsistent -- aren't these referring to python types, so this should be ComplexType?

SplitInfinity replied:
I was under the impression that there will eventually be a `ComplexFloatType` as well but if not, then I agree that it should be `ComplexType`.

anjali411 replied:
hmm I didn't know that we follow the python naming here. yeah should be `ComplexType`!

gchanan commented on torch/csrc/jit/serialization/unpickler.cpp:
```diff
@@ -543,6 +541,15 @@ void Unpickler::readGlobal(
     // Unpickle a tensor
     bool quantized = class_name == "_rebuild_qtensor";
     rebuildTensor(quantized);
+  } else if (module_name == "builtins" && class_name == "complex") {
```
above I see some stuff about floatlists -- we don't need to handle complexlists because we don't define them in e.g. native_functions yet?

gchanan replied:
it's also a little unclear to me if this is handling endianness -- do you know?

SplitInfinity replied:
The real and imaginary parts of the complex number are pickled using `pushIValue` which seems to take care of it?
```
void Pickler::pushDouble(double value) {
  push<PickleOpCode>(PickleOpCode::BINFLOAT);
  // Python pickle format is big endian, swap.
  push<double>(swapDouble(value));
}
```

```
double Unpickler::readFloat() {
  AT_ASSERT(sizeof(double) == 8);
  double big_endian = read<double>();
  double little_endian;

  // Pickle floats are big endian, so reverse the bytes
  auto big_endian_ptr = reinterpret_cast<const char*>(&big_endian);
  std::reverse_copy(
      big_endian_ptr,
      big_endian_ptr + sizeof(big_endian),
      reinterpret_cast<char*>(&little_endian));

  return little_endian;
}
```
